src/gdb-stub/gdb_stub.cpp

```diff
@@ -0,0 +1,492 @@
+            }
+        }
+
+        void read_registers(const connection_handler& connection, debugging_handler& handler)
+        {
+            std::string response{};
+            std::vector<std::byte> data{};
+            data.resize(handler.get_max_register_size());
+
+            const auto registers = handler.get_register_count();
+
+            for (size_t i = 0; i < registers; ++i)
+            {
+                memset(data.data(), 0, data.size());
+                const auto res = handler.read_register(i, data.data(), data.size());
+
+                if (!res)
+                {
+                    connection.send_reply("E01");
+                    return;
+                }
+
+                response.append(utils::string::to_hex_string(data));
+            }
+
+            connection.send_reply(response);
+        }
+
+        void write_registers(const connection_handler& connection, debugging_handler& handler,
+                             const std::string_view payload)
+        {
+            const auto data = utils::string::from_hex_string(payload);
+
+            const auto registers = handler.get_register_count();
+            const auto register_size = handler.get_max_register_size();
+
+            for (size_t i = 0; i < registers; ++i)
+            {
+                const auto offset = i * register_size;
+                const auto end_offset = offset + register_size;
+
+                if (data.size() < end_offset)
+                {
+                    connection.send_reply("E01");
+                    return;
+                }
+
+                const auto res = handler.write_register(i, data.data() + offset, register_size);
+
+                if (!res)
+                {
+                    connection.send_reply("E01");
+                    return;
+                }
+            }
+
+            connection.send_reply("OK");
+        }
+
+        void read_single_register(const connection_handler& connection, debugging_handler& handler,
+                                  const std::string& payload)
+        {
+            size_t reg{};
+            rt_assert(sscanf_s(payload.c_str(), "%zx", &reg) == 1);
+
+            std::vector<std::byte> data{};
+            data.resize(handler.get_max_register_size());
+
+            const auto res = handler.read_register(reg, data.data(), data.size());
+
+            if (res)
+            {
+                connection.send_reply(utils::string::to_hex_string(data));
+            }
+            else
+            {
+                connection.send_reply("E01");
+            }
+        }
+
+        void write_single_register(const connection_handler& connection, debugging_handler& handler,
+                                   const std::string_view payload)
+        {
+            const auto [reg, hex_data] = split_string(payload, '=');
+
+            size_t register_index{};
+            rt_assert(sscanf_s(std::string(reg).c_str(), "%zx", &register_index) == 1);
```

Comment:
probably this also. @0xeb maybe you can test this :)

Reply:
![image](https://github.com/user-attachments/assets/2ee441e7-774e-4479-8c68-1ab7ef212e56)

same for this. %zx should be fine